src/mpid/ch4/netmod/ofi/init_provider.c

```diff
@@ -0,0 +1,247 @@
+        return FI_VERSION(MPIDI_OFI_MAJOR_VERSION, MPIDI_OFI_MINOR_VERSION);
+    else
+        return fi_version();
+}
+
+/* find best matching provider */
+static int find_provider(struct fi_info **prov_out)
+{
+    int mpi_errno = MPI_SUCCESS;
+    int ret;                    /* return from fi_getinfo() */
+
+    const char *provname = MPIR_CVAR_OFI_USE_PROVIDER;
+    int required_version = MPIDI_OFI_get_required_version();
+    if (MPIR_CVAR_CH4_OFI_CAPABILITY_SETS_DEBUG && MPIR_Process.rank == 0) {
+        printf("Required minimum FI_VERSION: %x, current version: %x\n", required_version,
+               fi_version());
+    }
+
+    struct fi_info *hints = fi_allocinfo();
+    MPIR_Assert(hints != NULL);
+
+    if (MPIDI_OFI_ENABLE_RUNTIME_CHECKS) {
+        /* NOTE: prov_list should not be freed until we initialize multi-nic */
+        MPIDI_OFI_CALL(fi_getinfo(required_version, NULL, NULL, 0ULL, NULL, &prov_list), getinfo);
+
+        /* Pick a best matching provider and use it to refine hints */
+        struct fi_info *prov;
+        prov = pick_provider_from_list(provname, prov_list);
+
+        MPIR_ERR_CHKANDJUMP(prov == NULL, mpi_errno, MPI_ERR_OTHER, "**ofid_getinfo");
+
+        /* OFI provider doesn't expose FI_DIRECTED_RECV by default for performance consideration.
+         * MPICH should request this flag to enable it. */
+        if (prov->caps & FI_TAGGED) {
+            prov->caps |= FI_DIRECTED_RECV;
+        }
+
+        provname = prov->fabric_attr->prov_name;
+        /* Initialize MPIDI_OFI_global.settings */
+        MPIDI_OFI_init_settings(&MPIDI_OFI_global.settings, provname);
+        /* NOTE: we fill the hints with optimal settings. In order to work for providers
+         * that support halfway between minimal and optimal, we need try-loop to systematically
+         * relax settings. The following code does this for the providers that we have tested.
+         */
+        MPIDI_OFI_init_hints(hints);
+        hints->fabric_attr->prov_name = MPL_strdup(provname);
+        hints->caps = prov->caps;
+        hints->addr_format = prov->addr_format;
+
+        /* Now we have the hints with best matched provider, get the new prov_list */
+        struct fi_info *old_prov_list = prov_list;
+
+        /* First try with the initial (optimal) hints */
+        ret = fi_getinfo(required_version, NULL, NULL, 0ULL, hints, &prov_list);
+        if (ret || prov_list == NULL) {
+            /* relax msg_order */
+            hints->tx_attr->msg_order = prov->tx_attr->msg_order;
+            ret = fi_getinfo(required_version, NULL, NULL, 0ULL, hints, &prov_list);
+        }
+        if (ret || prov_list == NULL) {
+            if (prov->domain_attr->mr_mode & FI_MR_BASIC) {
+                hints->domain_attr->mr_mode = FI_MR_BASIC;
+            } else if (prov->domain_attr->mr_mode & FI_MR_SCALABLE) {
+                hints->domain_attr->mr_mode = FI_MR_SCALABLE;
+            }
```

Comment:
I'm not sure this makes sense anymore. `FI_MR_BASIC` and `FI_MR_SCALABLE` have been deprecated since OFI 1.5. `FI_MR_SCALABLE` is equivalent to `0x0` and `FI_MR_BASIC` is composed of a number of other mode bits that may or may not be set (e.g. `FI_MR_VIRT_ADDR`, `FI_MR_ALLOCATED`, and `FI_MR_PROV_KEY`).

Reply:
See the code in https://github.com/pmodels/mpich/pull/5233/files#diff-4a7220dc82d27330f373809888548c6250f7fa9316436874bafc7105f4638d21R124-R158 for details.

Reply:
Not all providers have moved away from `FI_MR_BASIC/FI_MR_SCALABLE`, especially those unmaintained providers such as `sockets`. If `fi_getinfo` return `mr_mode` with `FI_MR_BASIC` or `FI_MR_SCALABLE`, then we need set the hints using the legacy bit.

Reply:
https://github.com/ofiwg/libfabric/blob/14b3cd0450aebe8b74a2c5ed219a53cac06d3e0a/prov/sockets/src/sock_ep.c#L1450-L1457

The sockets provider only has a reference to scalable/basic to support backward compatibility. internally, it uses mode bits like all the rest.